src/backend/distributed/executor/multi_utility.c

```diff
@@ -964,13 +1031,153 @@ ErrorIfUnsupportedAlterTableStmt(AlterTableStmt *alterTableStatement)
+									errdetail("Citus cannot execute ADD CONSTRAINT "
+											  "command other than ADD CONSTRAINT FOREIGN "
+											  "KEY.")));
+				}
+
+				/*
+				 * ON DELETE SET NULL and ON DELETE SET DEFAULT is not supported. Because
+				 * we do not want to set partition column to NULL or default value.
+				 */
+				if (constraint->fk_del_action == FKCONSTR_ACTION_SETNULL ||
+					constraint->fk_del_action == FKCONSTR_ACTION_SETDEFAULT)
+				{
+					ereport(ERROR, (errcode(ERRCODE_FEATURE_NOT_SUPPORTED),
+									errmsg("cannot create foreign key constraint"),
+									errdetail("SET NULL or SET DEFAULT is not supported"
+											  " in ON DELETE operation.")));
+				}
+
+				/*
+				 * ON UPDATE SET NULL, ON UPDATE SET DEFAULT and UPDATE CASCADE is not
+				 * supported. Because we do not want to set partition column to NULL or
+				 * default value. Also cascading update operation would require
+				 * re-partitioning. Updating partition column value is not allowed anyway
+				 * even outside of foreign key concept.
+				 */
+				if (constraint->fk_upd_action == FKCONSTR_ACTION_SETNULL ||
+					constraint->fk_upd_action == FKCONSTR_ACTION_SETDEFAULT ||
+					constraint->fk_upd_action == FKCONSTR_ACTION_CASCADE)
+				{
+					ereport(ERROR, (errcode(ERRCODE_FEATURE_NOT_SUPPORTED),
+									errmsg("cannot create foreign key constraint"),
+									errdetail("SET NULL, SET DEFAULT or CASCADE is not"
+											  " supported in ON UPDATE operation.")));
+				}
+
+				/*
+				 * We will use constraint name in each placement by extending it at
+				 * workers. Therefore we require it to be exist.
+				 */
+				if (constraint->conname == NULL)
+				{
+					ereport(ERROR, (errcode(ERRCODE_FEATURE_NOT_SUPPORTED),
+									errmsg("cannot create foreign key constraint"),
+									errdetail("Creating foreign constraint without a "
+											  "name on a distributed table is currently "
+											  "not supported.")));
+				}
+
+				/* to enforce foreign constraints, tables must be co-located */
+				referencingTableId = RangeVarGetRelid(alterTableStatement->relation,
+													  lockmode,
+													  alterTableStatement->missing_ok);
+				referencedTableId = RangeVarGetRelid(constraint->pktable, lockmode,
+													 alterTableStatement->missing_ok);
+
+				if (!TablesColocated(referencingTableId, referencedTableId))
+				{
+					ereport(ERROR, (errcode(ERRCODE_FEATURE_NOT_SUPPORTED),
+									errmsg("cannot create foreign key constraint"),
+									errdetail("Foreign key constraint can only be created"
+											  " on co-located tables."),
+									errhint("If these tables are physically co-located"
+											" try mark them as co-located with"
+											" mark_tables_colocated() function.")));
+				}
+
+				/*
+				 * Referencing column's list length should be equal to referenced columns
+				 * list length.
+				 */
+				if (constraint->fk_attrs->length != constraint->pk_attrs->length)
```

Comment:
@byucesoy it is interesting that we've talked on this line and not realized the bug